### PR TITLE
[#903]: Make schema optional, allow alternate content type

### PR DESCRIPTION
**Pull request by ctron**
This change allows to make the schema optional and also allows to
configure the data content type.

Signed-off-by: Jens Reimann <jreimann@redhat.com>

### CODE REVIEW

**Comment by ctron** [comment]
Hm, looks like you have a CI check in place that doesn't allow to properly specify the copyright year range.

**Comment by thjaeckle** [comment]
> Hm, looks like you have a CI check in place that doesn't allow to properly specify the copyright year range.

Yes, copyright header for us is only the creation year as this is sufficient according to the Eclipse Foundation Handbook.

**Comment by ctron** [comment]
> > Hm, looks like you have a CI check in place that doesn't allow to properly specify the copyright year range.
> 
> Yes, copyright header for us is only the creation year as this is sufficient according to the Eclipse Foundation Handbook.

Works for me :+1:

File: services/gateway/endpoints/src/main/java/org/eclipse/ditto/services/gateway/endpoints/routes/cloudevents/CloudEventsRoute.java [hunk 1]
```diff
@@ -251,28 +260,52 @@ private CloudEvent validateCloudEvent(final CloudEvent cloudEvent, final Request
...
+     * @param dataContentType The content type to check.
+     * @return {@code true} if the content type is acceptable, {@code false} otherwise.
+     */
+    private boolean isCorrectDataType(@Nullable final String dataContentType) {
+        if (dataContentType == null) {
+            // no content type
+            return false;
+        }
+
+        final var result = MediaType.parse(dataContentType);
```
Comment (thjaeckle) [inline services/gateway/endpoints/src/main/java/org/eclipse/ditto/services/gateway/endpoints/routes/cloudevents/CloudEventsRoute.java:299]: Please use `akka.http.javadsl.model.ContentTypes.parse(String)` instead (whenever possible, use Akka's "javadsl" packages which contain the Java APIs - using the Scala APIs also work but often looks "strange").
The returned `ContentType` has a `mediaType()` accessor.
Please also don't use `var` keyword here - we did not yet adopt this ;)

File: services/gateway/util/src/main/java/org/eclipse/ditto/services/gateway/util/config/endpoints/DefaultCloudEventsConfig.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+import org.eclipse.ditto.services.utils.config.ConfigWithFallback;
+import org.eclipse.ditto.services.utils.config.ScopedConfig;
+
+import com.typesafe.config.Config;
+
+/**
+ * This class is the default implementation of the cloud events endpoint config.
+ */
+@Immutable
+public class DefaultCloudEventsConfig implements CloudEventsConfig {
```
Comment (thjaeckle) [inline services/gateway/util/src/main/java/org/eclipse/ditto/services/gateway/util/config/endpoints/DefaultCloudEventsConfig.java:29]: Please make this class `final`

File: services/gateway/util/src/main/java/org/eclipse/ditto/services/gateway/util/config/endpoints/DefaultCloudEventsConfig.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        DefaultCloudEventsConfig that = (DefaultCloudEventsConfig) o;
```
Comment (thjaeckle) [inline services/gateway/util/src/main/java/org/eclipse/ditto/services/gateway/util/config/endpoints/DefaultCloudEventsConfig.java:72]: please add `final`

File: services/gateway/util/src/main/java/org/eclipse/ditto/services/gateway/util/config/endpoints/DefaultCloudEventsConfig.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    @Override
+    public int hashCode() {
+        return Objects.hash(emptySchemaAllowed, dataTypes);
+    }
+
+    @Override
+    public String toString() {
+        return getClass().getSimpleName() + " [" +
+                "emptySchemaAllowed=" + emptySchemaAllowed +
+                "dataTypes=" + dataTypes +
```
Comment (thjaeckle) [inline services/gateway/util/src/main/java/org/eclipse/ditto/services/gateway/util/config/endpoints/DefaultCloudEventsConfig.java:85]:
```suggestion
                ", dataTypes=" + dataTypes +
```

File: services/gateway/util/src/main/java/org/eclipse/ditto/services/gateway/util/config/endpoints/CloudEventsConfig.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+        /**
+         * Flag if an empty data schema is allowed.
+         */
+        EMPTY_SCHEMA_ALLOWED("empty-schema-allowed", true),
+
+        /**
+         * Set of allowed data types
+         */
+        DATA_TYPES("data-types", Set.of("application/json", "application/vnd.eclipse.ditto+json"));
```
Comment (thjaeckle) [inline services/gateway/util/src/main/java/org/eclipse/ditto/services/gateway/util/config/endpoints/CloudEventsConfig.java:55]: Please use `org.eclipse.ditto.model.base.common.DittoConstants.DITTO_PROTOCOL_CONTENT_TYPE` for the Ditto Protocol content-type.
`application/json` could also be referenced e.g. via `MediaTypes.APPLICATION_JSON.toString()`.

File: services/gateway/util/src/main/java/org/eclipse/ditto/services/gateway/util/config/endpoints/DefaultCloudEventsConfig.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+        return emptySchemaAllowed;
+    }
+
+    @Override
+    public Set<String> getDataTypes() {
+        return dataTypes;
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
Comment (thjaeckle) [inline services/gateway/util/src/main/java/org/eclipse/ditto/services/gateway/util/config/endpoints/DefaultCloudEventsConfig.java:65]: please add `final` for `o`

**Review by thjaeckle — CHANGES_REQUESTED** [review]
Thx @ctron, already looks very promising :)
Added a few requests..

One more thing: 
Could you please enhance the `services/gateway/starter/src/main/resources/gateway.conf` with the `cloud-events` section (e.g. below `public-health { ... }`)  and the same defaults as in the code, so:
```
cloud-events {
  empty-schema-allowed = true
  data-types = [
    "application/json"
    "application/vnd.eclipse.ditto+json"
  ]
}
```
That way Ditto's configuration files act as documentation of what can be configured overall.

**Comment by ctron** [comment]
I just updated the PR.

File: services/gateway/endpoints/src/main/java/org/eclipse/ditto/services/gateway/endpoints/routes/cloudevents/CloudEventsRoute.java [hunk 7]
```diff
@@ -20,6 +20,7 @@
 
 import javax.annotation.Nullable;
 
+import akka.http.javadsl.model.ContentTypes;
```
Comment (thjaeckle) [inline services/gateway/endpoints/src/main/java/org/eclipse/ditto/services/gateway/endpoints/routes/cloudevents/CloudEventsRoute.java:23]: Double import can be removed

File: services/gateway/endpoints/src/test/java/org/eclipse/ditto/services/gateway/endpoints/routes/cloudevents/CloudEventsTest.java [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+import org.junit.Test;
+
+import java.util.Set;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public final class CloudEventsTest extends EndpointTestBase {
+
+    @Test
+    public void testDefaults() {
```
Comment (thjaeckle) [inline services/gateway/endpoints/src/test/java/org/eclipse/ditto/services/gateway/endpoints/routes/cloudevents/CloudEventsTest.java:27]: IMO this test does not make much sense as it just ensures what already is added as defaults in the `CloudEventsConfigValue`. If you enhance the default config, you must also enhance the test accordingly.

File: services/gateway/endpoints/src/test/java/org/eclipse/ditto/services/gateway/endpoints/routes/cloudevents/CloudEventsTest.java [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+import org.eclipse.ditto.services.gateway.endpoints.EndpointTestBase;
+import org.eclipse.ditto.services.gateway.util.config.endpoints.CloudEventsConfig;
+import org.junit.Test;
+
+import java.util.Set;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public final class CloudEventsTest extends EndpointTestBase {
```
Comment (thjaeckle) [inline services/gateway/endpoints/src/test/java/org/eclipse/ditto/services/gateway/endpoints/routes/cloudevents/CloudEventsTest.java:24]: What we normally add to our endpoint tests can be seen e.g. in: https://github.com/eclipse/ditto/blob/master/services/gateway/endpoints/src/test/java/org/eclipse/ditto/services/gateway/endpoints/routes/things/ThingsRouteTest.java

* build the route
* wrap it in Akka HTTP testkit's `testRoute` (as seen in `setUp()`)
* run `POST ` requests against the endpoint providing `ce-*` headers or omitting them in order to test for the exception which may be returned

**Review by thjaeckle — CHANGES_REQUESTED** [review]
The `CloudEventsTests` would need meaningful tests, but I'm not sure if you want to address all mentioned follow-up tasks of #903 ;)
I would rather delete the current `CloudEventsTests` class and the one not so useful test to be honest.

File: services/gateway/endpoints/src/test/java/org/eclipse/ditto/services/gateway/endpoints/routes/cloudevents/CloudEventsTest.java (hunk 8 above)
Comment (ctron) [reply services/gateway/endpoints/src/test/java/org/eclipse/ditto/services/gateway/endpoints/routes/cloudevents/CloudEventsTest.java:27]: Well it was for me, testing of the constants that I used actually provide the expected content. They do. But I thought I would leave the test in there, to ensure that is true in the future.

But I can also remove them if you prefer that.

File: services/gateway/endpoints/src/test/java/org/eclipse/ditto/services/gateway/endpoints/routes/cloudevents/CloudEventsTest.java (hunk 9 above)
Comment (ctron) [reply services/gateway/endpoints/src/test/java/org/eclipse/ditto/services/gateway/endpoints/routes/cloudevents/CloudEventsTest.java:24]: I can do that, but would tackle the problem in another PR.

**Comment by ctron** [comment]
> The `CloudEventsTests` would need meaningful tests, but I'm not sure if you want to address all mentioned follow-up tasks of #903 ;)
> I would rather delete the current `CloudEventsTests` class and the one not so useful test to be honest.

I wanted to keep the scope of this PR as is, if possible. And work on the other items in a new PR. Just to keep PRs small.

I tried to explain the idea behind the current state of the unit test above. But as mentioned, I am open to dropping it altogether for this PR.

File: services/gateway/endpoints/src/test/java/org/eclipse/ditto/services/gateway/endpoints/routes/cloudevents/CloudEventsTest.java (hunk 8 above)
Comment (thjaeckle) [reply services/gateway/endpoints/src/test/java/org/eclipse/ditto/services/gateway/endpoints/routes/cloudevents/CloudEventsTest.java:27]: Yes, I would prefer to delete that again 👍

**Review by thjaeckle — APPROVED** [review]

**Comment by thjaeckle** [comment]
Thx for the contribution, really appreciated 👍

**Pull request merged**